### PR TITLE
Rename `JSONToMessages` to `OpenAIToMessages`

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [x] other (please add here)

Closes #890. OpenAIToMessages is more recognizable and we can point to an actual reference for their chat completion APIs.

#### Changelog
What are the changes made in this PR?
* Change all instances of JSONToMessages to OpenAIToMessages and "json" to "openai"
* Support image content in openai messages
* move some utilities around for better dependency graph, avoiding cycles

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [x] I have added an example to docs or docstrings

### CODE REVIEW

**Review by felipemello1 — APPROVED** [review]

File: torchtune/data/_messages.py [hunk 1]
```diff
@@ -393,10 +393,10 @@ def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
         return {"messages": messages}
 
 
-class JSONToMessages(Transform):
+class OpenAIToMessages(Transform):
     """
-    Convert a single chat sample with identical json structure to torchtune's :class:`~torchtune.data.Message`
-    structure. This transform simply creates Message dataclasses from the provided jsons.
+    Convert a single chat sample adhering to the `OpenAI chat completion <https://platform.openai.com/docs/api-reference/chat>`_
+    json structure to torchtune's :class:`~torchtune.data.Message` structure.
```
Comment (SalmanMohammadi) [inline torchtune/data/_messages.py:399]:
```suggestion
    JSON structure to torchtune's :class:`~torchtune.data.Message` structure.
```

Comment (RdoubleA) [reply torchtune/data/_messages.py:399]: at this point it should be a real word in the dictionary

**Review by SalmanMohammadi — APPROVED** [review]

**Review by joecummings — CHANGES_REQUESTED** [review]
checking something don't merge yet

**Comment by joecummings** [comment]
Okay, yeah there are two different ways that OpenAI allows inputs:

1. Just text input:
```
curl https://api.openai.com/v1/chat/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer $OPENAI_API_KEY" \
  -d '{
    "model": "gpt-4o",
    "messages": [
      {
        "role": "system",
        "content": "You are a helpful assistant."
      },
      {
        "role": "user",
        "content": "Hello!"
      }
    ]
  }'
```

2. Multimodal input:
```
curl https://api.openai.com/v1/chat/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer $OPENAI_API_KEY" \
  -d '{
    "model": "gpt-4o",
    "messages": [
      {
        "role": "user",
        "content": [
          {
            "type": "text",
            "text": "What'\''s in this image?"
          },
          {
            "type": "image_url",
            "image_url": {
              "url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg"
            }
          }
        ]
      }
    ],
    "max_tokens": 300
  }'
```

The current method supports ONLY the first situation.

**Review by joecummings — APPROVED** [review]
i love it

**Pull request merged**